File: engine/rules/conditions.py
"""Condition mutation helpers."""

from __future__ import annotations

from typing import Dict, Iterable


def normalize_condition_name(name: str) -> str:
    return name.lower().replace(" ", "_")


def condition_is_immune(name: str, condition_immunities: Iterable[str]) -> bool:
    normalized = normalize_condition_name(name)
    immunity_set = {normalize_condition_name(x) for x in condition_immunities}
    return normalized in immunity_set or "all_conditions" in immunity_set


def apply_condition(conditions: Dict[str, int], name: str, value: int = 1) -> Dict[str, int]:
    result = dict(conditions)
    key = normalize_condition_name(name)
    current = result.get(key, 0)
    result[key] = max(current, value)
    return result


def clear_condition(conditions: Dict[str, int], name: str) -> Dict[str, int]:
    result = dict(conditions)
    key = normalize_condition_name(name)
    result.pop(key, None)
    return result

File: src/rules/conditions.ts
/**
 * Condition mutation helpers.
 * Mirrors engine/rules/conditions.py
 *
 * Pathfinder 2e ORC conditions: tracked as {name: value} where value is
 * usually a numeric severity (e.g. frightened 2).
 */

export function normalizeConditionName(name: string): string {
  return name.toLowerCase().replace(/ /g, "_");
}

export function conditionIsImmune(
  name: string,
  conditionImmunities: string[],
): boolean {
  const normalized = normalizeConditionName(name);
  const immunitySet = new Set(conditionImmunities.map(normalizeConditionName));
  return immunitySet.has(normalized) || immunitySet.has("all_conditions");
}

export function applyCondition(
  conditions: Record<string, number>,
  name: string,
  value = 1,
): Record<string, number> {
  const result = { ...conditions };
  const key = normalizeConditionName(name);
  const current = result[key] ?? 0;
  result[key] = Math.max(current, value);
  return result;
}

export function clearCondition(
  conditions: Record<string, number>,
  name: string,
): Record<string, number> {
  const result = { ...conditions };
  const key = normalizeConditionName(name);
  delete result[key];
  return result;
}
